### PR TITLE
NNI search fixes

### PR DESCRIPTION
**Pull request by davidrich27**
## Description

- Fixed some issues with NNI search routine.
- Fixed issue with UpdateChoiceMap causing slowdown.
- Fixed issue with Update routine that sometimes ran on inactive ParsimonyEngine.
- Added method for generating Top Tree Newick for TPEngine and Spanning Newick for SubsplitDAG.
- Added timer wrapper for function.
- Added some print functions for DAGStorage.

## Tests

- Added TEST_CASE("TPEngine: Exporting Newicks")
  - Run multiple NNI iterations, builds Newicks, reconstructs fresh TPEngine from resulting Newick, then checks that TPEngines match.

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [x] Issue branch has been squashed and rebased on main branch
* [ ] GitHub CI build on PR branch completed successfully

### CODE REVIEW

File: src/generic_tree_collection.hpp [hunk 1]
```diff
@@ -51,6 +51,8 @@ class GenericTreeCollection {
   size_t TreeCount() const { return trees_.size(); }
   const TTreeVector &Trees() const { return trees_; }
   const TTree &GetTree(size_t i) const { return trees_.at(i); }
+  // A tag is a packed int of two values: (1) node id, (2) leaf count below node.  For
+  // taxons, the node id is the taxon id and the leaf count is 1.
```
Comment (matsen) [inline src/generic_tree_collection.hpp:55]:
```suggestion
  // taxa, the node id is the taxon id and the leaf count is 1.
```

File: src/gp_doctest.cpp [hunk 2]
```diff
@@ -2760,6 +2760,188 @@ TEST_CASE("TPEngine: Branch Length Optimization") {
...
+  nni_engine.SetTopNScoreFilteringScheme(1);
+  nni_engine.SetReevaluateRejectedNNIs(true);
+  nni_engine.RunInit(true);
+
+  for (size_t iter = 0; iter < 5; iter++) {
+    nni_engine.RunMainLoop(true);
+    // nni_engine.GetDAG().PrintStorage();
+
+    // if (iter == 7) break;
+    // Issue #xxx: this creates an unknown problem on iteration 7.
```
Comment (matsen) [inline src/gp_doctest.cpp:2925]: Let's make an issue if this is a problem...

File: src/gp_instance.cpp [hunk 3]
```diff
@@ -741,14 +767,36 @@ void GPInstance::SubsplitDAGToDot(const std::string &out_path,
   out_stream.close();
 }
 
+void GPInstance::ExportSpanningTreesWithGPBranchLengths(
```
Comment (matsen) [inline src/gp_instance.cpp:770]: Is this our first use of the word "spanning trees"? 

Spanning tree has a meaning https://en.wikipedia.org/wiki/Spanning_tree

Is the notion that it's a set of trees such that they generate the whole DAG? Perhaps "CoveringTrees"? We can discuss.

File: src/subsplit_dag_storage.hpp [hunk 4]
```diff
@@ -569,6 +569,12 @@ class SubsplitDAGStorage {
     if (!result.has_value()) {
       return {};
     }
+    // if (std::get<0>(result.value()).value_ >= lines_.size()) {
```
Comment (matsen) [inline src/subsplit_dag_storage.hpp:572]: cruft?

**Review by matsen — COMMENTED** [review]
A fairly surface-level pass... but a few things I'd like to address.

File: src/generic_tree_collection.hpp (hunk 1 above)
Comment (davidrich27) [reply src/generic_tree_collection.hpp]: Fixed.

File: src/gp_doctest.cpp (hunk 2 above)
Comment (davidrich27) [reply src/gp_doctest.cpp]: Added issue.

File: src/gp_instance.cpp (hunk 3 above)
Comment (davidrich27) [reply src/gp_instance.cpp]: This is the first usage of the term.  Yes, the goal is to create a Newick can generate the DAG with all edges and vertices, with a minimum set of trees.  I agree, covering trees does sound like a better name for this.

File: src/gp_doctest.cpp [hunk 5]
```diff
@@ -2760,6 +2760,188 @@ TEST_CASE("TPEngine: Branch Length Optimization") {
...
+      GPInstanceOfFiles(fasta_path, newick_path_2, "_ignore/mmapped_pv.data2");
+  inst_1.MakeTPEngine();
+  inst_1.MakeNNIEngine();
+  auto& tp_engine = inst_1.GetTPEngine();
+  auto& nni_engine = inst_1.GetNNIEngine();
+  inst_2.MakeTPEngine();
+  inst_2.MakeNNIEngine();
+
+  CHECK_MESSAGE(
+      TPEngine::Compare(inst_1.GetTPEngine(), inst_1.GetTPEngine(), false) == 0,
```
Comment (matsen) [inline src/gp_doctest.cpp:2894]: Do you mean to have inst_1 and inst_2 here?

Comment (matsen) [reply src/gp_doctest.cpp:2894]: Or are you actually testing the Compare functionality?

**Review by matsen — APPROVED** [review]
Looks great, Dave! Just two little things here. Let's get this in!

File: src/gp_doctest.cpp (hunk 5 above)
Comment (davidrich27) [reply src/gp_doctest.cpp:2894]: Yes it is testing compare functionality.

**Pull request merged**